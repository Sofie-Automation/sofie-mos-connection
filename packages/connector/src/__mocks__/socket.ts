/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Socket } from 'net'

import * as iconv from 'iconv-lite'
import { getMessageId } from '../__tests__/lib.js'
import { getMosTypes } from '@mos-connection/model'
import { Mocked, vitest } from 'vitest'
iconv.encodingExists('utf16-be')

const mosTypes = getMosTypes(true)

// Mock the Socket class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<ISocketMock> = []

export class SocketMock extends EventEmitter implements Socket {
	// Mock implementation:
	bufferSize = 0
	remoteAddress = ''
	remoteFamily = ''
	remotePort = 0
	localAddress = ''
	localPort = 0
	bytesRead = 0
	bytesWritten = 0
	connecting = false
	destroyed = false
	writable = false
	readable = false

	readonly readableEncoding: BufferEncoding | null = null
	readonly readableEnded: boolean = false
	readonly readableFlowing: boolean | null = null

	readonly writableEnded: boolean = false
	readonly writableFinished: boolean = false
	readonly writableHighWaterMark: number = 0
	readonly writableLength: number = 0
	readonly writableObjectMode: boolean = false

	readonly readableHighWaterMark: number = 0
	readonly readableLength: number = 0
	readonly readableObjectMode: boolean = false

	public name = ''
	public connectedPort = 0
	public connectedHost = ''

	// [Symbol.asyncIterator](): AsyncIterableIterator<any>

	private _responses: Array<ReplyTypes> = []
	private _autoReplyToHeartBeat = true
	public mockConnectCount = 0

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		// Wrap member functions in mocks:
		this.connect = vitest.fn(this.connect)
		this.setEncoding = vitest.fn(this.setEncoding)
		this.destroy = vitest.fn(this.destroy)
		this.pause = vitest.fn(this.pause)
		this.resume = vitest.fn(this.resume)
		this.setTimeout = vitest.fn(this.setTimeout)
		this.setNoDelay = vitest.fn(this.setNoDelay)
		this.setKeepAlive = vitest.fn(this.setKeepAlive)
		this.address = vitest.fn(this.address)
		this.unref = vitest.fn(this.unref)
		this.ref = vitest.fn(this.ref)
		this.end = vitest.fn(this.end)
		this._write = vitest.fn(this._write)
		this.setDefaultEncoding = vitest.fn(this.setDefaultEncoding)
		this._read = vitest.fn(this._read)
		this.read = vitest.fn(this.read)
		this.isPaused = vitest.fn(this.isPaused)
		this.pipe = vitest.fn(this.pipe) as any
		this.unpipe = vitest.fn(this.unpipe)
		this.unshift = vitest.fn(this.unshift)
		this.wrap = vitest.fn(this.wrap)
		this.push = vitest.fn(this.push)
		this._destroy = vitest.fn(this._destroy)
		this._final = vitest.fn(this._final)
		this.cork = vitest.fn(this.cork)
		this.uncork = vitest.fn(this.uncork)

		this.mockSentMessage0 = vitest.fn(this.mockSentMessage0)
		this.mockSentMessage = vitest.fn(this.mockSentMessage)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): ISocketMock[] {
		return instances
	}

	write(data: unknown, encoding: unknown): boolean {
		this.mockSentMessage0(data, encoding as any)
		return true
	}
	// @ts-expect-error mock
	connect(port: number, host: string): this {
		this.mockConnectCount++
		this.connectedPort = port
		this.connectedHost = host

		return this
	}
	setEncoding(): this {
		return this
	}
	destroy(_error?: Error): this {
		this.destroyed = true
		this.emit('close')
		/* nothing */
		return this
	}
	pause(): this {
		return this
	}
	resume(): this {
		return this
	}
	setTimeout(timeout: number, callback?: (...args: any[]) => void): this {
		if (callback) setTimeout(callback, timeout)
		return this
	}
	setNoDelay(_noDelay?: boolean): this {
		return this
	}
	setKeepAlive(_enable?: boolean, _initialDelay?: number): this {
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 100, family: 'localhost', address: '127.0.0.1' }
	}
	unref(): this {
		return this
	}
	ref(): this {
		return this
	}
	end(): this {
		return this
	}
	_write(_chunk: unknown, _encoding: string, callback: () => void): void {
		callback() /* nothing */
	}
	setDefaultEncoding(_encoding: string): this {
		return this
	}
	_read(_size: number): void {
		// nothing
	}
	read(_size?: number): void {
		// nothing
	}
	isPaused(): boolean {
		return false
	}
	pipe<T extends NodeJS.WritableStream>(destination: T, _options?: { end?: boolean }): T {
		return destination
	}
	unpipe(_destination: NodeJS.WritableStream): this {
		return this
	}
	unshift(_chunk: unknown): void {
		// nothing
	}
	wrap(_oldStream: NodeJS.ReadableStream): this {
		return this
	}
	push(_chunk: unknown, _encoding?: string): boolean {
		return true
	}
	_destroy(): void {
		/* nothing */
	}
	_final(): void {
		/* nothing */
	}
	cork(): void {
		/* nothing */
	}
	uncork(): void {
		/* nothing */
	}

	// ------------------------------------------------------------------------
	// Mock methods:
	mockEmitConnected(): void {
		if (this.connectedPort === 10542) {
			// don't reply on heartbeats on query port
			this._autoReplyToHeartBeat = false
		}

		this.emit('connect')
	}
	mockEmitClose(): void {
		this.emit('close')
	}

	mockSentMessage0(data: unknown, encoding: string): void {
		if (this._autoReplyToHeartBeat) {
			const str: string = typeof data === 'string' ? data : this.decode(data as any)

			if (/<heartbeat>/.exec(str)) {
				try {
					const mosID = (str.match(/<mosID>([^<]+)<\/mosID>/) || [])[1]
					const ncsID = (str.match(/<ncsID>([^<]+)<\/ncsID>/) || [])[1]
					const messageId = getMessageId(str)

					const repl = `<mos>
<mosID>${mosID}</mosID>
<ncsID>${ncsID}</ncsID>
<messageID>${messageId}</messageID>\
  <heartbeat>
    <time>${mosTypes.mosTime.stringify(mosTypes.mosTime.create(Date.now()))}</time>
  </heartbeat>
</mos>\r\n`
					this.mockReceiveMessage(this.encode(repl))
				} catch (e) {
					// eslint-disable-next-line no-console
					console.error('mockReply', str)
					throw e
				}
				return
			}
		}
		this.mockSentMessage(data, encoding)
	}
	mockSentMessage(data: unknown, _encoding: string): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			setTimeoutOrg(() => {
				Promise.resolve(typeof cb === 'function' ? cb(data) : cb)
					.then((msg) => {
						if (msg) {
							if (Array.isArray(msg)) {
								for (const m of msg) {
									this.mockReceiveMessage(m)
								}
							} else {
								this.mockReceiveMessage(msg)
							}
						}
					})
					// eslint-disable-next-line no-console
					.catch(console.error)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: ReplyTypes): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		void this._responses.splice(0, 9999)
		// @ts-expect-error mock hack
		this.mockSentMessage0['mockClear']()
		// @ts-expect-error mock hack
		this.mockSentMessage['mockClear']()
	}
	async mockWaitForSentMessages(): Promise<void> {
		return new Promise<void>((resolve) => {
			const check = () => {
				if (this._responses.length === 0) {
					resolve()
				} else {
					setTimeoutOrg(() => {
						check()
					}, 1)
				}
			}
			check()
		})
	}

	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
	setAutoReplyToHeartBeat(autoReplyToHeartBeat: boolean): void {
		this._autoReplyToHeartBeat = autoReplyToHeartBeat
	}
	get autoReplyToHeartBeat(): boolean {
		return this._autoReplyToHeartBeat
	}
}

type SimpleTypes = string | string[] | false | Buffer | Buffer[]

type ReplyTypes =
	| SimpleTypes
	| ((data: any) => SimpleTypes)
	| Promise<SimpleTypes>
	| ((data: any) => Promise<SimpleTypes>)

export type ISocketMock = Mocked<SocketMock>
