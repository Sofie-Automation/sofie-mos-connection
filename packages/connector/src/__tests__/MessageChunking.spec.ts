import {
	checkMessageSnapshot,
	clearMocks,
	decode,
	doBeforeAll,
	encode,
	getMessageId,
	getMosConnection,
	getMosDevice,
	getXMLReply,
	makeFakeIncomingMessage,
	mosTypes,
	sendFakeIncomingMessage,
	setupMocks,
} from './lib.js'
import {
	IMOSListMachInfo,
	IMOSObject,
	IMOSROAck,
	IMOSROFullStory,
	IMOSRunningOrder,
	MosConnection,
	MosDevice,
} from '../index.js'
import { SocketMock } from '../__mocks__/socket.js'
import { xmlData, xmlApiData } from '../__mocks__/testData.js'

beforeAll(() => {
	setupMocks()
})
beforeEach(() => {
	clearMocks()
})

describe('message chunking', () => {
	let mosDevice: MosDevice
	let mosConnection: MosConnection
	let socketMockLower: SocketMock
	let socketMockUpper: SocketMock

	let serverSocketMockLower: SocketMock
	let serverSocketMockUpper: SocketMock
	let serverSocketMockQuery: SocketMock

	let onRequestMachineInfo: jest.Mock<any, any>
	let onRequestMOSObject: jest.Mock<any, any>
	let onRequestAllMOSObjects: jest.Mock<any, any>

	let onRunningOrderStory: jest.Mock<any, any>

	beforeAll(async () => {
		SocketMock.mockClear()
		// ServerMock.mockClear()

		mosConnection = await getMosConnection(
			{
				'0': true,
				'1': true, // Must support at least one other profile
			},
			false
		)
		mosDevice = await getMosDevice(mosConnection)

		// Profile 0:
		onRequestMachineInfo = jest.fn(async () => {
			return xmlApiData.machineInfo
		})
		mosDevice.onRequestMachineInfo(async (): Promise<IMOSListMachInfo> => {
			return onRequestMachineInfo()
		})
		// Profile 1:
		onRequestMOSObject = jest.fn(async () => {
			return xmlApiData.mosObj
		})
		onRequestAllMOSObjects = jest.fn(async () => {
			return [xmlApiData.mosObj, xmlApiData.mosObj2]
		})
		mosDevice.onRequestMOSObject(async (objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})
		mosDevice.onRequestAllMOSObjects(async (): Promise<Array<IMOSObject>> => {
			return onRequestAllMOSObjects()
		})

		const roAckReply = async () => {
			const ack: IMOSROAck = {
				ID: mosTypes.mosString128.create('runningOrderId'),
				Status: mosTypes.mosString128.create('OK'),
				Stories: [],
			}
			return ack
		}
		// Profile 3:
		onRunningOrderStory = jest.fn(roAckReply)

		mosDevice.onRunningOrderStory(async (story: IMOSROFullStory): Promise<IMOSROAck> => {
			return onRunningOrderStory(story)
		})
		const b = doBeforeAll()
		socketMockLower = b.socketMockLower
		socketMockUpper = b.socketMockUpper

		serverSocketMockLower = b.serverSocketMockLower
		serverSocketMockUpper = b.serverSocketMockUpper
		serverSocketMockQuery = b.serverSocketMockQuery
	})
	afterAll(async () => {
		await mosConnection.dispose()
	})
	beforeEach(() => {
		onRunningOrderStory.mockClear()

		serverSocketMockLower.mockClear()
		serverSocketMockUpper.mockClear()
		serverSocketMockQuery.mockClear()
	})
	test('init', async () => {
		expect(mosDevice).toBeTruthy()
		expect(socketMockLower).toBeTruthy()
		expect(socketMockUpper).toBeTruthy()
	})

	function chunkSubstr(str: string, size: number) {
		const numChunks = Math.ceil(str.length / size)
		const chunks = new Array(numChunks)

		/* tslint:disable-next-line */
		for (let i = 0, o = 0; i < numChunks; ++i, o += size) {
			chunks[i] = str.slice(o, o + size)
		}

		return chunks
	}

	test('chunks', async () => {
		// Prepare server response
		const mockReply = jest.fn((data: Buffer) => {
			const str = decode(data)
			const messageID = getMessageId(str)
			const repl = getXMLReply(messageID, xmlData.roList)
			const chunks = chunkSubstr(repl, 500)
			expect(chunks).toHaveLength(4)
			return chunks.map((c) => encode(c))
		})

		socketMockUpper.mockAddReply(mockReply)
		if (!xmlApiData.roList.ID) throw new Error(`xmlApiData.roList.ID not set`)
		const returnedObj = (await mosDevice.sendRequestRunningOrder(xmlApiData.roList.ID)) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		const msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()
	})

	test('chunk around space', async () => {
		// Prepare server response
		const mockReply = jest.fn((data) => {
			const str = decode(data)
			const messageID = getMessageId(str)
			const repl = getXMLReply(messageID, xmlData.roList)

			const splitPoint = repl.indexOf('Test MOS')
			expect(splitPoint).not.toEqual(-1)

			const chunks = [repl.slice(0, splitPoint + 4), repl.slice(splitPoint + 4)]
			return chunks.map((c) => encode(c))
		})

		socketMockUpper.mockAddReply(mockReply)
		if (!xmlApiData.roList.ID) throw new Error(`xmlApiData.roList.ID not set`)
		const returnedObj = (await mosDevice.sendRequestRunningOrder(xmlApiData.roList.ID)) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		const msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()

		// This is what we split in half
		const storySlug = mosTypes.mosString128.create(returnedObj.Stories[1].Slug)
		expect(storySlug).toBeTruthy()
		expect(mosTypes.mosString128.stringify(storySlug)).toEqual('Test MOS')
	})

	test('junk data before', async () => {
		// Prepare server response
		const mockReply = jest.fn((data) => {
			const str = decode(data)
			const messageID = getMessageId(str)
			const repl = getXMLReply(messageID, xmlData.roList)

			const padded = '         JUNK DATA   ' + repl
			return encode(padded)
		})

		socketMockUpper.mockAddReply(mockReply)
		if (!xmlApiData.roList.ID) throw new Error(`xmlApiData.roList.ID not set`)
		const returnedObj = (await mosDevice.sendRequestRunningOrder(xmlApiData.roList.ID)) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		const msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()
	})

	test('junk data packet before', async () => {
		// Prepare server response
		const mockReply = jest.fn((data) => {
			const str = decode(data)
			const messageID = getMessageId(str)
			const repl = getXMLReply(messageID, xmlData.roList)

			const chunks = ['         JUNK DATA   ', repl]
			return chunks.map((c) => encode(c))
		})

		socketMockUpper.mockAddReply(mockReply)
		if (!xmlApiData.roList.ID) throw new Error(`xmlApiData.roList.ID not set`)
		const returnedObj = (await mosDevice.sendRequestRunningOrder(xmlApiData.roList.ID)) as IMOSRunningOrder
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		const msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReq>/)
		checkMessageSnapshot(msg)
		expect(returnedObj).toMatchObject(xmlApiData.roList2)
		expect(returnedObj).toMatchSnapshot()
	})

	test('incoming chunked message', async () => {
		const onRequestMOSObject = jest.fn(async (_objId: string) => {
			return xmlApiData.mosObj
		})
		onRequestMOSObject.mockClear()
		mosDevice.onRequestMOSObject(async (objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})

		const message = makeFakeIncomingMessage(`<mosReqObj><objID>M000123</objID></mosReqObj>`)

		const chunks = [message.message.slice(0, 100), message.message.slice(100)]

		// Send first part of the message:
		sendFakeIncomingMessage(serverSocketMockLower, chunks[0])
		expect(onRequestMOSObject).toHaveBeenCalledTimes(0)

		// Send rest of the message:
		sendFakeIncomingMessage(serverSocketMockLower, chunks[1])
		expect(onRequestMOSObject).toHaveBeenCalledTimes(1)
	})

	test('multiple mos tags', async () => {
		const onRequestMOSObject = jest.fn(async (_objId: string) => {
			return xmlApiData.mosObj
		})
		onRequestMOSObject.mockClear()
		mosDevice.onRequestMOSObject(async (objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})

		const message = makeFakeIncomingMessage(
			`<mosReqObj><objID>M000123</objID><mos><test>hehehe</test></mos></mosReqObj>`
		)

		const i0 = message.message.indexOf('<mos>', 10)
		const i1 = message.message.indexOf('</mos>', i0 + 1)
		const i2 = message.message.indexOf('</mos>', i1 + 1)

		const chunks = [
			message.message.slice(0, i0),
			message.message.slice(i0, i1),
			message.message.slice(i1, i2),
			message.message.slice(i2),
		]

		// Send the parts of the message:
		sendFakeIncomingMessage(serverSocketMockLower, chunks[0])
		expect(onRequestMOSObject).toHaveBeenCalledTimes(0)
		sendFakeIncomingMessage(serverSocketMockLower, chunks[1])
		expect(onRequestMOSObject).toHaveBeenCalledTimes(0)
		sendFakeIncomingMessage(serverSocketMockLower, chunks[2])
		expect(onRequestMOSObject).toHaveBeenCalledTimes(0)
		sendFakeIncomingMessage(serverSocketMockLower, chunks[2])
		expect(onRequestMOSObject).toHaveBeenCalledTimes(1)
	})
	test('multiple messages', async () => {
		const onRequestMOSObject = jest.fn(async (_objId: string) => {
			return xmlApiData.mosObj
		})
		onRequestMOSObject.mockClear()
		mosDevice.onRequestMOSObject(async (objId: string): Promise<IMOSObject | null> => {
			return onRequestMOSObject(objId)
		})

		const message0 = makeFakeIncomingMessage(
			`<mosReqObj><objID>M000123</objID><mos><test>hehehe</test></mos></mosReqObj>`
		)
		const message1 = makeFakeIncomingMessage(
			`<mosReqObj><objID>M000124</objID><mos><test>hahaha</test></mos></mosReqObj>`
		)

		// Send both messages right away:
		sendFakeIncomingMessage(serverSocketMockLower, message0.message + message1.message)
		expect(onRequestMOSObject).toHaveBeenCalledTimes(2)
	})
})
