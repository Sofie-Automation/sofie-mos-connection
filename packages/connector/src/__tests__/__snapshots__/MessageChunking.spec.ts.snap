// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`message chunking chunk around space 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking chunk around space 2`] = `
{
  "ID": {
    "_mosString128": "96857485",
  },
  "Slug": {
    "_mosString128": "5PM RUNDOWN",
  },
  "Stories": [
    {
      "ID": {
        "_mosString128": "5983A501:0049B924:8390EF2B",
      },
      "Items": [
        {
          "EditorialDuration": 645,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000224",
          },
          "Paths": [
            {
              "Description": "MPEG2 Video",
              "Target": "\\server\\media\\clip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            {
              "Description": "WM9 750Kbps",
              "Target": "https://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            {
              "Description": "MOS Object",
              "Target": "https://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": {
            "_mosString128": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B10",
      },
      "Slug": {
        "_mosString128": "Colstat Murder",
      },
    },
    {
      "ID": {
        "_mosString128": "3854737F:0003A34D:983A0B28",
      },
      "Items": [
        {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B11",
      },
      "Slug": {
        "_mosString128": "Test MOS",
      },
    },
  ],
}
`;

exports[`message chunking chunks 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking chunks 2`] = `
{
  "ID": {
    "_mosString128": "96857485",
  },
  "Slug": {
    "_mosString128": "5PM RUNDOWN",
  },
  "Stories": [
    {
      "ID": {
        "_mosString128": "5983A501:0049B924:8390EF2B",
      },
      "Items": [
        {
          "EditorialDuration": 645,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000224",
          },
          "Paths": [
            {
              "Description": "MPEG2 Video",
              "Target": "\\server\\media\\clip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            {
              "Description": "WM9 750Kbps",
              "Target": "https://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            {
              "Description": "MOS Object",
              "Target": "https://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": {
            "_mosString128": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B10",
      },
      "Slug": {
        "_mosString128": "Colstat Murder",
      },
    },
    {
      "ID": {
        "_mosString128": "3854737F:0003A34D:983A0B28",
      },
      "Items": [
        {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B11",
      },
      "Slug": {
        "_mosString128": "Test MOS",
      },
    },
  ],
}
`;

exports[`message chunking junk data before 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking junk data before 2`] = `
{
  "ID": {
    "_mosString128": "96857485",
  },
  "Slug": {
    "_mosString128": "5PM RUNDOWN",
  },
  "Stories": [
    {
      "ID": {
        "_mosString128": "5983A501:0049B924:8390EF2B",
      },
      "Items": [
        {
          "EditorialDuration": 645,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000224",
          },
          "Paths": [
            {
              "Description": "MPEG2 Video",
              "Target": "\\server\\media\\clip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            {
              "Description": "WM9 750Kbps",
              "Target": "https://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            {
              "Description": "MOS Object",
              "Target": "https://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": {
            "_mosString128": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B10",
      },
      "Slug": {
        "_mosString128": "Colstat Murder",
      },
    },
    {
      "ID": {
        "_mosString128": "3854737F:0003A34D:983A0B28",
      },
      "Items": [
        {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B11",
      },
      "Slug": {
        "_mosString128": "Test MOS",
      },
    },
  ],
}
`;

exports[`message chunking junk data packet before 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReq>
    <roID>M000123</roID>
  </roReq>
</mos>"
`;

exports[`message chunking junk data packet before 2`] = `
{
  "ID": {
    "_mosString128": "96857485",
  },
  "Slug": {
    "_mosString128": "5PM RUNDOWN",
  },
  "Stories": [
    {
      "ID": {
        "_mosString128": "5983A501:0049B924:8390EF2B",
      },
      "Items": [
        {
          "EditorialDuration": 645,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000224",
          },
          "Paths": [
            {
              "Description": "MPEG2 Video",
              "Target": "\\server\\media\\clip392028cd2320s0d.mxf",
              "Type": "PATH",
            },
            {
              "Description": "WM9 750Kbps",
              "Target": "https://server/proxy/clipe.wmv",
              "Type": "PROXY PATH",
            },
            {
              "Description": "MOS Object",
              "Target": "https://server/proxy/clipe.xml",
              "Type": "METADATA PATH",
            },
          ],
          "Slug": {
            "_mosString128": "COLSTAT MURDER:VO",
          },
          "Trigger": "CHAINED",
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B10",
      },
      "Slug": {
        "_mosString128": "Colstat Murder",
      },
    },
    {
      "ID": {
        "_mosString128": "3854737F:0003A34D:983A0B28",
      },
      "Items": [
        {
          "EditorialDuration": 310,
          "EditorialStart": 55,
          "ID": {
            "_mosString128": "0",
          },
          "MOSID": "testmos.enps.com",
          "MosExternalMetaData": [
            {
              "MosPayload": {
                "Owner": "SHOLMES",
                "destination": "b",
                "source": "a",
                "transitionMode": "2",
                "transitionPoint": "463",
              },
              "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
              "MosScope": "PLAYLIST",
            },
          ],
          "ObjectID": {
            "_mosString128": "M000133",
          },
          "UserTimingDuration": 310,
        },
      ],
      "Number": {
        "_mosString128": "B11",
      },
      "Slug": {
        "_mosString128": "Test MOS",
      },
    },
  ],
}
`;
